### PR TITLE
Integrate Tasks API

### PR DESCRIPTION
**Pull request by Giost**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] PR has a descriptive title
- [x] PR contains closely related changes only (should not have many unrelated changes)
- [x] STEP pod members have been added as reviewers
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #14 

## What is the new behavior?

Integrate the Tasks API providing an class for retrieving task list, tasks and updating a task.
Clean `CalendarClientHelperTest`.

### CODE REVIEW

**Review by fuiszgt — APPROVED** [review]
LGTM.

File: src/main/java/com/google/sps/api/tasks/TasksInterface.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  public List<Task> getTasks(String tasksListID) throws IOException {
+    List<Task> tasks = tasksClient.tasks().list(tasksListID).execute().getItems();
+    return TasksHelper.filterTasks(tasks);
+  }
+
+  /**
+   * Returns the tasks belonging to the most recently updated task list.
+   */
+  public List<Task> getTasksMostRecentList() throws IOException {
```
Comment (dola) [inline src/main/java/com/google/sps/api/tasks/TasksInterface.java:63]: `getTasksOfMostRecentList`
names are long anyway. So you can as well leave a preposition in there.

File: src/main/java/com/google/sps/api/tasks/TasksInterface.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    return getTasks(TasksHelper.getIdMostRecentTaskList(getTasksList()));
+  }
+
+  /**
+   * Returns the specified task belonging to the task list specified.
+   */
+  public Task getTask(String tasksListID, String taskID) throws IOException {
```
Comment (dola) [inline src/main/java/com/google/sps/api/tasks/TasksInterface.java:70]: also name it `Id` not `ID`

Comment (dola) [reply src/main/java/com/google/sps/api/tasks/TasksInterface.java:70]: same a few more times below

File: src/main/java/com/google/sps/api/tasks/TasksInterface.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  public void updateTask(String taskListID, Task task) throws IOException {
+    tasksClient.tasks().update(taskListID, task.getId(), task).execute();
+  }
+
+  /**
+   * Updates the date of the specified task.
+   * The date is specified using a RFC 3339 timestamp.
+   */
+  public void updateDateTimeTask(String taskListID, String taskID, String dateRfc3339) throws IOException {
```
Comment (dola) [inline src/main/java/com/google/sps/api/tasks/TasksInterface.java:85]: while you can name it dateRfc3339 it doesn't help too much in this name here. I'd rather mention it in the JavaDoc similar to how date libaries do it. (using a @param line in the doc)

File: src/main/java/com/google/sps/api/tasks/TasksHelper.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * Provides some helper methods for tasks.
+ */
+public class TasksHelper {
+  /**
+   * Returns a list of tasks containing only those without a date or past their due date.
+   */
+  public static List<Task> filterTasks(List<Task> tasks) {
+    DateTime dateTimeNow = new DateTime(
+        new Date());
+    long epochNow = dateTimeNow.getValue();
```
[the remark this answers is not in the document]
Comment (Giost) [reply src/main/java/com/google/sps/api/tasks/TasksHelper.java:36]: I think it's the same

File: src/main/java/com/google/sps/api/tasks/TasksHelper.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        .filter(task -> task.getDue() == null || TimeConverter.dateToEpoch(task.getDue()) < epochNow)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Returns the task list ID of the most recent updated list.
+   */
+  public static String getIdMostRecentTaskList(List<TaskList> taskList) {
+    return taskList.stream()
+        .max(Comparator.comparingLong(tasksList -> TimeConverter.dateToEpoch(tasksList.getUpdated())))
```
[the remark this answers is not in the document]
Comment (Giost) [reply src/main/java/com/google/sps/api/tasks/TasksHelper.java:48]: Yeah, I was also surprised.

File: src/main/java/com/google/sps/api/tasks/TasksInterface.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+import com.google.api.services.tasks.model.TaskLists;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.sps.api.authorization.AuthorizationRequester;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Provides access to the Tasks API.
+ */
+public class TasksInterface {
```
[the remark this answers is not in the document]
Comment (Giost) [reply src/main/java/com/google/sps/api/tasks/TasksInterface.java:31]: I agree.

**Review by dola — APPROVED** [review]
Cool! Did you use the IntelliJ refactorings to rename variables and classes? That would make it work in a few simple steps rather than having to do search and replace.

**Comment by Giost** [comment]
> Cool! Did you use the IntelliJ refactorings to rename variables and classes? That would make it work in a few simple steps rather than having to do search and replace.

Absolutely, I used that and it is really easier.

**Pull request merged**